oeps/architectural-decisions/oep-0058-arch-translations-management.rst

```diff
@@ -0,0 +1,263 @@
+order to grant the app permissions to push/pull the translation files. Currently, we
+manage the push/pull permissions for the edx-transifex-bot through a number of GitHub
+user groups. The `Transifex GitHub App`_ once installed in an organization, is granted
+permissions to push/pull on a repository basis, and by moving all the translation files
+to a single repository we eliminate separate translations user groups.
+
+Connect the New Translation Repository to Transifex
+===================================================
+
+The Transifex web-app accepts a `Transifex GitHub Integration configuration file`_ for
+each Transifex project. By connecting the single repository containing all translation
+files, we only need to make a single `Transifex GitHub Integration configuration file`_
+that allows the `Transifex GitHub App`_ to manage the translation files. Based on the
+Translation Working Group's instruction on acceptable translation/review percentages, we
+can set parameters that automatically push and pull translation files.
+
+Copy Transifex's Translation Memory and Combine Translators
+===========================================================
+
+As a last step we will reorganize the openedx Transifex organization by combining
+translators and reviewers across Transifex projects into the new project associated with
+the new repository. In addition, we can save all the progress the Open edX translators
+have accomplished by copying the Transifex's Translation Memory, the auto-translation
+feature that allows for Projects with similar strings to be automatically translated,
+from the old projects to this new one. Once older projects are made redundant by the new
+project, they will be deprecated. By moving all the translation files to the same
+repository we can increase the reach of the Transifex's Translation Memory feature to
+help translate similar strings across the entire code/documentation base.
+
+Get Translations Back for Deployment/Development
+================================================
+
+A new python library, called openedx-atlas, will be created. This will enable the
+placement of the translation files kept in openedx-translations into locally cloned
+repositories for development and containers containing the code translation files are
+formed from. This tool will manage the placement of translation files through an editable
+atlas configuration file (atlas.yml) kept in the repositories that have
+translation files kept in openedx-translations. The atlas.yml file will support
+options that allow for the concatenation, reorganization, and reformatting of translation
+files as they are copied to their locations amongst the code. The atlas.yml file
+will also support selecting which languages to be included in an Open edX deployment. The
+tool will have to be used/ran as part of the setup of a repository, whether for
+development or deployment.
```

Comment:
Since openedx-atlas works as a configurable wrapper to git's sparse-checkout, openedx-atlas can be used to pull translations from openedx-translations right after git clone is used to pull the repository from GitHub. This should cover most supported development & deployment environments, though I'm curious if there are any others that may not work like this.

Reply:
> This should cover most supported development & deployment environments, though I'm curious if there are any others that may not work like this.

Having atlas built in python adds complexity to the deployment of MFEs. It seems there are a few paths that we could go down

The first option would be to keep atlas in python, and to use the python implementation everywhere
```
Pros:
* Flexibile/Extensibile
* Easy to maintain (only need to maintain one implementation)
Cons:
* Requires python in build/deployment envs
```

Another option would be to write atlas in bash, that way python would not be required in build/deployment environments for MFEs.

```
Pros:
* Runs (pretty much) everywhere
Cons:
* Bash can get messy when projects get complex
```

A third option would be to *also* write atlas functionality in js and package for NPM so MFEs (or frontend-build) could import it.

```
Pros:
* Easy integration for MFEs
* Flexible/extensible
Cons:
* Requires maintaining duplicated functionality (need to implement everything in both js and python)
```

If the scope of atlas stays limited, i'd lean towards the bash option. If there's an expectation of increased complexity within atlas, however, there are definite benefits to using full fledged programming languages.

Reply:
@bmedx mentioned a fourth option: get rid of openedx-atlas, and publish releases of .tar.gz/.zip files from the openedx-translations repository
```
Pros:
* No sparse checkout logic needed, no separate CLI tool needed.
* All the complexity would be in the openedx-translations repo; other repos would just download & unpack the .tar.gzs.
Cons:
* Would need to add automation to the openedx-translations repo to regularly publish these
* The .tar.gzs would contain every language (or every supported language?)
  * It's unclear whether the .tar.gzs would be too big or not.
     edx-platform's locale directory is currently 60 Mb, gzips down to 14 Mb.
* Is there other stuff we think openedx-atlas would do that we'd be leaving out with this option?
```

Reply:
Solely from the MFE repo perspective, I'm for gzips.  Github workflows for publishing them on every commit would be pretty simple.  And I very much doubt download sizes will be problematic.  For all I know `tx pull` could be even less efficient.  🤷🏼 

What would stop us from storing translations per-repo in opendx-translations, though?  That would make each individual pull pretty tiny.

If that doesn't work, for whatever reason, I'd vote for bash.

Reply:
@arbrandes 

> What would stop us from storing translations per-repo in opendx-translations, though? That would make each individual pull pretty tiny.

That is indeed the suggestion on the table -- one gzip published per repo 🙂 It's just that each gzip would have every language *unless* we wanted to publish R*L gzips, where R is the number of repos and L is the number of languages.